### DISCUSSION TITLE
How can I set App title in winui3 desktop app

### DISCUSSION THREAD

**Discussion by ApoorvaPetkar** [microsoft/WindowsAppSDK#2994] [Q&A]
I followed this link on setting the title which got closed with no clear solution - https://github.com/microsoft/WindowsAppSDK/issues/1914

I'm still getting cast exceptions. Can you please post a proper working screenshot?

Can someone help how I can achieve this?

**Comment by ghost1372** [comment]
In App.cs you can set title before activate window:

```
m_window = new MainWindow();
m_window.Title = "test";
m_window.Activate();
```
Or if you want to use AppWindow:

```
public MainWindow()
 {
 this.InitializeComponent(); 
m_AppWindow = GetAppWindowForCurrentWindow();
 m_AppWindow.Title = "App title"; 
} 

private AppWindow GetAppWindowForCurrentWindow() 
{ 
IntPtr hWnd = WindowNative.GetWindowHandle(this);
 WindowId wndId = Win32Interop.GetWindowIdFromWindow(hWnd); 
return AppWindow.GetFromWindowId(wndId);
 }
```

**Comment by DarranRowe** [comment]
I think the thread you referenced is clear enough.
WindowNative wrapper is backed by the [IWindowNative](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/win32/microsoft.ui.xaml.window/nn-microsoft-ui-xaml-window-iwindownative) interface. This explicitly states:
"This interface is implemented by Microsoft.UI.Xaml.Window, which desktop apps can use to get the underlying HWND of the window."
So, for the default WinUI 3 project layout, this has to be executed on the MainWindow object. If you notice, in the thread you reference, the cast exception happens on objects which derive from Page.

**Reply by DarranRowe** [reply]
To show this in action.

![Screenshot 2022-09-28 133914](https://user-images.githubusercontent.com/52577874/192781087-7a204649-9529-4d0e-aaf2-734fc9dbf97f.png)

![Screenshot 2022-09-28 134045](https://user-images.githubusercontent.com/52577874/192781134-df09f4e9-583e-4b71-92ca-740120a3f3a8.png)

This stores the AppWindow instance in the class and then uses that to set the title. But as you should be able to see from the screenshots, this works without issue. But also notice how this is happening in the constructor for a class that derives from Microsoft.UI.Xaml.Window.
